processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java

```diff
@@ -122,14 +122,13 @@ public Assignment getTargetAssignment(Method mappingMethod, String mappedElement
         ResolvingAttempt attempt = new ResolvingAttempt(
             sourceModel,
             mappingMethod,
-            mappedElement,
             dateFormat,
             numberFormat,
             sourceRHS,
             criteria
         );
 
-        return attempt.getTargetAssignment( sourceRHS.getSourceType(), targetType );
+        return attempt.getTargetAssignment( sourceRHS.getSourceTypeForMatching(), targetType );
```

Comment:
I'll add: `The source type that is to be used when resolving the mapping from source to target.` WDYT? We stay "in terminology" using terms resolving and mapping.